include/oneapi/dpl/internal/binary_search_impl.h

```diff
@@ -164,13 +184,30 @@ upper_bound_impl(__internal::__hetero_tag<_BackendTag>, Policy&& policy, InputIt
     auto keep_result = oneapi::dpl::__ranges::__get_sycl_range<__bknd::access_mode::read_write, OutputIterator>();
     auto result_buf = keep_result(result, result + value_size);
     auto zip_vw = make_zip_view(input_buf.all_view(), value_buf.all_view(), result_buf.all_view());
-    __bknd::__parallel_for(_BackendTag{}, ::std::forward<Policy>(policy),
-                           custom_brick<StrictWeakOrdering, decltype(size), upper_bound>{comp, size}, value_size,
-                           zip_vw)
-        .wait();
+    // Enable index calculation to proceed with uint32_t if input range is small enough.
+    if (size <= ::std::numeric_limits<::std::uint32_t>::max())
+    {
+        __bknd::__parallel_for(_BackendTag{}, ::std::forward<Policy>(policy),
```

Comment:
Seems like we can reduce this to a lambda and make the branches just provide `policy` and `size` of the appropriate type.

It adds the complexity of the lambda to reduce the code replication, so its not a clear win, but I think it would be better.
(same comment for all three instances)

Reply:
Changed for all three instances to use a lambda.